File: docs/templates/geocode-bigdatacloud-batch.py
#!/usr/bin/env python3
"""Reverse geocode via BigDataCloud (curl) — ưu tiên lat/lng cho trạm V2/thiếu huyện."""
import json, math, subprocess, time, re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
OUT = Path(__file__).resolve().parent / "geocode-unmapped-results.json"
STATIONS_TS = ROOT / "pages/station-coverage/stationCoverageFromDb.ts"

def plottable(lat, lng):
    if lat is None or lng is None: return False
    if not (8.35 <= lat <= 23.45 and 102.12 <= lng <= 109.55): return False
    if abs(lat - round(lat)) < 1e-6 and abs(lng - round(lng)) < 1e-6: return False
    return True

def load_stations():
    text = STATIONS_TS.read_text()
    marker = "export const DB_STATIONS"
    assign = text.index(marker)
    start = text.index("\n[", assign)
    end = text.index("] as DbStationRow", start)
    return json.loads(text[start:end + 1])

def bdc(lat, lon):
    url = f"https://api.bigdatacloud.net/data/reverse-geocode-client?latitude={lat}&longitude={lon}&localityLanguage=vi"
    # Dùng proxy Cursor khi có; retry khi 403
    last_err = None
    for attempt in range(4):
        r = subprocess.run(
            ["curl", "-sS", "--max-time", "25", url],
            capture_output=True,
            text=True,
        )
        if r.returncode == 0 and r.stdout.strip().startswith("{"):
            return json.loads(r.stdout)
        last_err = (r.stderr or r.stdout or "curl failed").strip()
        time.sleep(1.2 * (attempt + 1))
    raise RuntimeError(last_err or "curl failed")

def tokens_from(geo):
    toks = []
    def push(field, raw):
        if raw: toks.append(f"{field}={raw}")
    push("locality", geo.get("locality"))
    push("city", geo.get("city"))
    push("province", geo.get("principalSubdivision"))
    push("country", geo.get("countryName"))
    for a in (geo.get("localityInfo") or {}).get("administrative") or []:
        lvl = a.get("adminLevel") or 0
        name = a.get("name")
        if name and lvl >= 5:
            push(f"admin{lvl}", name)
        elif name and lvl > 3:
            push("admin", name)
    return toks

def main():
    existing = []
    if OUT.exists():
        existing = [r for r in json.loads(OUT.read_text()) if r.get("display_name") and not r.get("error")]
    have = {str(r["id"]) for r in existing}
    stations = load_stations()
    todo = []
    for s in stations:
        if str(s["id"]) in have: continue
        if not plottable(s.get("latitude"), s.get("longitude")): continue
        d = (s.get("districtCode") or "").strip()
        p = (s.get("precinctCode") or "").strip()
        if (not d) or re.fullmatch(r"\d{1,2}", d) or re.fullmatch(r"\d{5}", p):
            todo.append(s)
    print(f"Have OK: {len(existing)}; todo: {len(todo)}", flush=True)
    results = list(existing)
    for i, s in enumerate(todo, 1):
        print(f"[{i}/{len(todo)}] {s.get('code')} ... ", end="", flush=True)
        try:
            geo = bdc(s["latitude"], s["longitude"])
            cc = (geo.get("countryCode") or "").lower()
            display = ", ".join(filter(None, [geo.get("locality"), geo.get("city"), geo.get("principalSubdivision"), geo.get("countryName")]))
            results.append({
                "id": s["id"],
                "code": s.get("code"),
                "address": s.get("address"),
                "display_name": display,
                "tokens": tokens_from(geo),
                "country_code": cc,
                "preferredProvince": s.get("provinceCode"),
                "provider": "bigdatacloud",
            })
            print(f"{cc or '?'} · {geo.get('locality') or '-'} / {geo.get('city') or '-'}", flush=True)
        except Exception as e:
            results.append({"id": s["id"], "code": s.get("code"), "error": str(e), "provider": "bigdatacloud"})
            print(f"ERR {e}", flush=True)
            time.sleep(1.5)
        time.sleep(0.15)
        if i % 100 == 0:
            OUT.write_text(json.dumps(results, ensure_ascii=False, indent=2))
            print(f"checkpoint {len(results)}", flush=True)
    OUT.write_text(json.dumps(results, ensure_ascii=False, indent=2))
    ok = sum(1 for r in results if r.get("display_name") and not r.get("error"))
    err = sum(1 for r in results if r.get("error"))
    print(json.dumps({"total": len(results), "ok": ok, "err": err}, ensure_ascii=False), flush=True)

if __name__ == "__main__":
    main()

File: docs/templates/geocode-bigdatacloud-batch.mjs
/**
 * Reverse geocode (BigDataCloud) cho trạm plottable VN + mã huyện/xã kiểu V2,
 * ưu tiên lat/lng khi khớp master. Bỏ ngoài VN / không có master.
 *
 * Chạy: node docs/templates/geocode-bigdatacloud-batch.mjs
 * Sau đó: node docs/templates/merge-overrides-prefer-latlng.mjs
 */
import { readFileSync, writeFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const designRoot = join(__dirname, '../..');
const outPath = join(__dirname, 'geocode-unmapped-results.json');

function isPlottableVn(lat, lng) {
  if (lat == null || lng == null || !Number.isFinite(lat) || !Number.isFinite(lng)) return false;
  if (lat < 8.35 || lat > 23.45 || lng < 102.12 || lng > 109.55) return false;
  if (Math.abs(lat - Math.round(lat)) < 1e-6 && Math.abs(lng - Math.round(lng)) < 1e-6) return false;
  return true;
}

function loadStations() {
  const src = readFileSync(join(designRoot, 'pages/station-coverage/stationCoverageFromDb.ts'), 'utf8');
  const start = src.indexOf('[', src.indexOf('export const DB_STATIONS'));
  const end = src.indexOf('] as DbStationRow', start);
  return Function(`"use strict"; return (${src.slice(start, end + 1)});`)();
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

async function reverseBdc(lat, lon) {
  const url = new URL('https://api.bigdatacloud.net/data/reverse-geocode-client');
  url.searchParams.set('latitude', String(lat));
  url.searchParams.set('longitude', String(lon));
  url.searchParams.set('localityLanguage', 'vi');
  const res = await fetch(url);
  if (!res.ok) throw new Error(`BDC ${res.status}`);
  return res.json();
}

function tokensFromBdc(geo) {
  const tokens = [];
  const push = (field, raw) => {
    if (raw) tokens.push(`${field}=${raw}`);
  };
  push('locality', geo.locality);
  push('city', geo.city);
  push('province', geo.principalSubdivision);
  push('country', geo.countryName);
  const admins = geo.localityInfo?.administrative || [];
  for (const a of admins) {
    if (a.adminLevel >= 5 && a.adminLevel <= 8) push(`admin${a.adminLevel}`, a.name);
    else if (a.name && a.adminLevel > 3) push('admin', a.name);
  }
  return tokens;
}

async function main() {
  // Keep successful Nominatim rows; drop fetch-failed stubs
  let existing = [];
  try {
    existing = JSON.parse(readFileSync(outPath, 'utf8')).filter((r) => r.display_name && !r.error);
  } catch {
    existing = [];
  }
  const have = new Set(existing.map((r) => String(r.id)));
  const stations = loadStations();

  const todo = stations.filter((s) => {
    if (have.has(String(s.id))) return false;
    if (!isPlottableVn(s.latitude, s.longitude)) return false;
    const d = (s.districtCode || '').trim();
    const p = (s.precinctCode || '').trim();
    return !d || /^\d{1,2}$/.test(d) || /^\d{5}$/.test(p);
  });

  console.log(`Have OK: ${existing.length}; todo BDC: ${todo.length}`);
  const results = [...existing];

  for (let i = 0; i < todo.length; i += 1) {
    const s = todo[i];
    process.stdout.write(`[${i + 1}/${todo.length}] ${s.code} ... `);
    try {
      const geo = await reverseBdc(s.latitude, s.longitude);
      const cc = (geo.countryCode || '').toLowerCase();
      results.push({
        id: s.id,
        code: s.code,
        address: s.address,
        display_name: [geo.locality, geo.city, geo.principalSubdivision, geo.countryName]
          .filter(Boolean)
          .join(', '),
        tokens: tokensFromBdc(geo),
        country_code: cc,
        preferredProvince: s.provinceCode,
        provider: 'bigdatacloud',
      });
      console.log(`${cc || '?'} · ${geo.locality || '-'} / ${geo.city || '-'}`);
    } catch (e) {
      results.push({ id: s.id, code: s.code, error: String(e.message || e), provider: 'bigdatacloud' });
      console.log('ERR', e.message || e);
      await sleep(2000);
    }
    await sleep(200);
    if ((i + 1) % 100 === 0) {
      writeFileSync(outPath, JSON.stringify(results, null, 2));
      console.log(`checkpoint ${results.length}`);
    }
  }

  writeFileSync(outPath, JSON.stringify(results, null, 2));
  console.log(
    JSON.stringify(
      {
        total: results.length,
        ok: results.filter((r) => r.display_name && !r.error).length,
        err: results.filter((r) => r.error).length,
      },
      null,
      2,
    ),
  );
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
